onnxruntime/core/graph/model.h

```diff
@@ -165,13 +165,33 @@ class Model {
   // Get model's serialization proto data.
   ONNX_NAMESPACE::ModelProto ToProto();
 
+  // Get model's serialization proto data.
+  // Floating-point initializers are saved in the given external file.
+  // This function is useful to avoid hitting the size limit for protobuf files.
+  ONNX_NAMESPACE::ModelProto ToGraphProtoWithExternalInitializers(const std::string& external_file_name);
+
 #ifdef _WIN32
   static common::Status Save(Model& model, const std::wstring& file_path);
 #endif
   static common::Status Save(Model& model, const std::string& file_path);
 
   static common::Status Save(Model& model, int fd);
 
+  // Save the model to file using an external file for floating-point initializers.
+  // Notice that when on Windows the external_file_name is a plain string.
+  // This is because the string is saved inside the output protobuf as a plain string, where wchar is not supported.
+#ifdef _WIN32
+  static common::Status SaveWithExternalInitializers(Model& model,
+                                                     const std::wstring& file_path,
+                                                     const std::string& external_file_name);
+#else
+  static common::Status SaveWithExternalInitializers(Model& model,
+                                                     const std::string& file_path,
+                                                     const std::string& external_file_name);
+#endif
+
+  static common::Status SaveWithExternalInitializers(Model& model, int fd, const std::string& external_file_name);
+
   static common::Status Load(std::istream& model_istream, ONNX_NAMESPACE::ModelProto* p_model_proto);
```

Comment:
Not sure we need these static methods. Couldn't the ONNX external data helper script be used if the model has already been successfully saved?

Reply:
The problem with using the python script is that it assumes that dumping the full ONNX file succeeds. This is not the case when dealing with large models, for which the ONNX file exceeds the 2GB protobuf limit.

Reply:
What's the use case when you'd call this static method though? Isn't it dependent on a successful dump of the ONNX model?

Reply:
The use case is to dump large models after autodiff. Currently when running the bert training example the function Model::Save fails because the model hits the 2GB protobuf limit. The initializers are the largest component of the model. 
The function Model::SaveWithExternalInitializers can now be used in TrainingSession to avoid hitting the limit, since the graph itself (ignoring the initializers) is pretty small.